api/envoy/api/v2/route/route.proto

```diff
@@ -893,6 +893,20 @@ message RetryPolicy {
     google.protobuf.Duration max_interval = 2 [(validate.rules).duration = {gt {}}];
   }
 
+  message RetryBudget {
+    // Specifies the limit on concurrent retries as a percentage of outstanding requests. This
+    // parameter is optional.
+    //
+    // Defaults to 20%.
+    type.Percent percent_budget = 1;
+
+    // Specifies the minimum retry concurrency allowed for the retry budget. This parameter is
+    // optional.
+    //
+    // Defaults to 10.
+    google.protobuf.UInt32Value min_concurrency = 2;
```

Comment:
Does this mean, if based on percent budget , if 15 retries are allowed, only 10 retries will happen and 5 will wait?

Reply:
Looking at the code, looks like this concurrency is applied across active requests + retries?

Reply:
> Does this mean, if based on percent budget , if 15 retries are allowed, only 10 retries will happen and 5 will wait?

The budget is enforced after all other considerations for whether to retry. If `num_retries` is set to 15 and you are under that amount, the budget will be enforced if there are active request/connections that exceed the `min_concurrency`. The minimum concurrency is just a concurrency threshold at which we begin evaluating the budget, not a required number of  active retries.

> Looking at the code, looks like this concurrency is applied across active requests + retries?

Yes, but I'm open to discussion here. I made that decision because I thought of a scenario where multiple downstreams had a mix of HTTP/1.1 and HTTP/2 protocols with the upstream cluster. However, I could be misguided.

Reply:
Oh. OK. I thought specifying `num_retries` is not required and if retry budget is specified, it is considered that the user is interested in retries and retry will happen as long as you are in the budget and concurrent is set-  But looks like we need to specify num_retries as well? Of course we can give a high number in num_retries and then budget will enforce that request is retried if budget is there - Is that the intention?

Reply:
I had spoken about something like this with @mattklein123 and I think there's promise in allowing any number of retries and relying on the global timeout. There should be a bit more discussion around how to allow this though and I wanted to keep this patch minimal.

I would really like to allow this though, perhaps in a subsequent patch.

Reply:
Great. IMO, the retry budget driving how many retries can be made considering the budget + concurrency+current request rate without having to give fixed `num_retries` would be super useful.